### ISSUE TITLE
Loosen ParameterValue dtype/backend contract via Array-API Protocol

### ISSUE THREAD

**Issue by jc-macdonald** [ACCIDDA/flepimop2#270]
## Environment

- OS: macOS 15
- Python version: 3.13
- `flepimop2` version: `0.3.0.dev0` (current `main`)

## Reproducible example

```python
import jax, jax.numpy as jnp
from flepimop2.axis import ResolvedShape
from flepimop2.parameter.abc import ParameterValue

shape = ResolvedShape(axis_names=("draw",), sizes=(3,))

# Concrete jax.Array — silently round-trips through host memory, losing
# device placement / dtype / PyTree identity.
pv = ParameterValue(value=jnp.asarray([1.0, 2.0, 3.0]), shape=shape)
print(type(pv.value).__module__)  # 'numpy'

# jax.vmap tracer — hard error.
def make(x):
    return ParameterValue(value=jnp.broadcast_to(x, (3,)), shape=shape)
jax.vmap(make)(jnp.asarray([1.0, 2.0, 3.0]))
# jax.errors.TracerArrayConversionError
```

The offending line is `ParameterValue.__post_init__`:

```python
value = np.asarray(self.value, dtype=np.float64)
```

## Brief explanation

`ParameterValue` is a `(value, shape)` record. The shape contract is real
(`ResolvedShape` is checked); the dtype/backend contract — `value:
Float64NDArray` plus the unconditional `np.asarray(..., dtype=np.float64)`
cast — is over-specification. The cast is purely the runtime enforcement
of the static `Float64NDArray` annotation, and it has three real costs:

1. It silently coerces malformed plugin input (`int` arrays, `float32`
   arrays, Python lists) to `float64`, hiding bugs that the static
   type-check already catches.
2. It blocks any non-NumPy array backend from passing through. A concrete
   `jax.Array` survives via host round-trip (correct, but loses device
   placement and zero-copy semantics); a `jax.vmap` / `jit` tracer raises
   `TracerArrayConversionError` outright.
3. It implicitly couples `flepimop2` to the NumPy backend at a layer
   (`ParameterABC`'s return contract) where backend-agnosticism would
   otherwise be free.

The natural fix is the [Python Array API standard][1]: every modern array
library (NumPy ≥ 2.0, JAX ≥ 0.4.32, PyTorch ≥ 2.1, CuPy, dask) implements
`__array_namespace__`. A small `Protocol` in `flepimop2.typing` captures
the contract `ParameterValue` actually needs (`.shape` + array-API
membership) without taking a JAX/PyTorch/etc. dependency:

```python
# flepimop2/typing.py
from typing import Protocol, runtime_checkable

@runtime_checkable
class Array(Protocol):
    """Minimal Array-API duck type used by ``ParameterValue``."""
    @property
    def shape(self) -> tuple[int, ...]: ...
    @property
    def dtype(self) -> object: ...
    def __array_namespace__(
        self, *, api_version: str | None = None
    ) -> object: ...
```

```python
# flepimop2/parameter/abc/__init__.py
@dataclass(frozen=True, slots=True)
class ParameterValue:
    value: Array
    shape: ResolvedShape

    def __post_init__(self) -> None:
        if tuple(self.value.shape) != self.shape.sizes:
            raise ValueError(...)
```

That's the entire change. Producers continue to hand in
`numpy.ndarray` (which satisfies `Array`); JAX-driven producers hand in
`jax.Array` (which also satisfies `Array`); the engine's downstream
`jnp.asarray(...)` is unchanged. Plugins remain responsible for handing
in the correct dtype — which `mypy` already enforces.

### Backwards compatibility

- All existing `ParameterValue(np.asarray(...))` callers in
  `src/flepimop2/_cli/_simulate_command.py`,
  `src/flepimop2/parameter/fixed/__init__.py`, and the test suite are
  unaffected: NumPy arrays satisfy `Array`.
- The shape check is preserved verbatim.
- No third-party dependency added (the `Protocol` is duck-typed; nothing
  imports `jax`, etc.).

### Concrete downstream motivation

In ACCIDDA/flepimop2-demo we have a `numpyro_predictive` process plugin
that draws prior samples via `numpyro.infer.Predictive` and threads them
into a parametric expression via a `ContextVar` substitution. The demo's
plugins are already dual-path (numpy ↔ `jax.numpy`); the only place the
JAX path dies is the `ParameterValue` boundary. With this fix:

- `numpyro_predictive` becomes `jax.vmap(run_one)(prior_samples)` instead
  of N serial JIT'd calls.
- A future `numpyro_inference` plugin (NUTS / SVI) can wrap the engine in
  a single `numpyro` model with an observed-data likelihood — the
  gradient/tracer flow is currently blocked at this exact line.

Happy to send the PR (small — typing additions + the `__post_init__`
trim + a couple of unit tests).

[1]: https://data-apis.org/array-api/latest/
